bindings/spt/sys_linux_aarch64.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2015-2018 Contributors as noted in the AUTHORS file
+ *
+ * This file is part of Solo5, a sandboxed execution environment.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "bindings.h"
+
+/*
+ * TODO: Argument types need a re-check against the kernel types.
+ * We are most certainly missing clobbers in the syscall functions.
+ */
+
+#define SYS_read          63
+#define SYS_write         64
+#define SYS_pread64       67
+#define SYS_pwrite64      68
+#define SYS_clock_gettime 113
+#define SYS_exit_group    94
+#define SYS_ppoll         73
+
+int sys_read(int fd, char *buf, size_t size)
+{
+    register long x8 __asm__("x8") = SYS_read;
+    register long x0 __asm__("x0") = fd;
+    register long x1 __asm__("x1") = (long)buf;
+    register long x2 __asm__("x2") = (long)size;
+
+    __asm__ __volatile__ (
```

Comment:
I guess for `clock_gettime` we could skip the context switch entirely and just read from the vDSO, but it just occurred to me that your reason for not exposing those to the guest could be that it seems likely vvar/vdso could contain "sensitive" data belonging to the tender, like PID or cpu node or UID/GID or whatever ([looks like](http://man7.org/linux/man-pages/man7/vdso.7.html) that is platform dependent).

Reply:
I just used the convention that musl uses for aarch64. Isn't the `vsyscall` stuff ARM32 only? As for vDSO, yes, that has various issues associated with it, would prefer to avoid it for now.

Reply:
My linux-fu turned out to be very outdated; `vsyscall` is the ancient pre-ASLR method, now probably only found on old stuff.
They split that up into an r/w section `vvar` and an executable `vdso` for data and syscall wrappers respectively.
But if vDSO is a pain and this is easier then fair enough. :)